### PR TITLE
OSDP: Add support for Control Panel mode of operation

### PR DESCRIPTION
**Pull request by sidcha**
This is a follow-up patch (to #26965) add support for CP mode and a sample.

CC: @carlescufi @jukkar @pfalcon @galak

### CODE REVIEW

File: subsys/mgmt/osdp/src/osdp_cp.c [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+#define REPLY_LSTATR_DATA_LEN          2
+#define REPLY_RSTATR_DATA_LEN          1
+#define REPLY_COM_DATA_LEN             5
+#define REPLY_NAK_DATA_LEN             1
+#define REPLY_KEYPPAD_DATA_LEN         2   /* variable length command */
+#define REPLY_RAW_DATA_LEN             4   /* variable length command */
+#define REPLY_FMT_DATA_LEN             3   /* variable length command */
+#define REPLY_BUSY_DATA_LEN            0
+
+static int osdp_atoi(const char *a, int *i)
```
Comment (jukkar) [inline subsys/mgmt/osdp/src/osdp_cp.c:44]: I suggest that you use `strtol()` instead of this function, as strtol will handle errors just fine.

File: subsys/mgmt/osdp/src/osdp_cp.c [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+	if (a == NULL)
+		return -1;
+	val = atoi(a);
+	if (val == 0 && a[0] != '0')
+		return -1;
+	*i = val;
+	return 0;
+}
+
+static char *osdp_strsep(char **str, char *sep)
```
Comment (jukkar) [inline subsys/mgmt/osdp/src/osdp_cp.c:57]: Could you separate this function to minimal libc (send a separate PR for that) as it might be useful to other applications too.

File: subsys/mgmt/osdp/src/osdp_cp.c [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+static int cp_process_reply(struct osdp_pd *pd)
+{
+	uint8_t *buf;
+	int rec_bytes, ret, max_len;
+
+	buf = pd->rx_buf + pd->rx_buf_len;
+	max_len = sizeof(pd->rx_buf) - pd->rx_buf_len;
+
+	rec_bytes = pd->channel.recv(pd->channel.data, buf, max_len);
+	if (rec_bytes <= 0)	/* No data received */
```
Comment (jukkar) [inline subsys/mgmt/osdp/src/osdp_cp.c:507]: one liners need {}

File: subsys/mgmt/osdp/src/osdp_cp.c [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+	return (ret == len) ? 0 : -1;
+}
+
+/**
+ * cp_process_reply - received buffer from serial stream handling partials
+ * Returns:
+ *  0: success
+ * -1: error
+ *  1: no data yet
+ *  2: re-issue command
```
Comment (jukkar) [inline subsys/mgmt/osdp/src/osdp_cp.c:496]: Could you use defines for the return values as it is easier for the reader. Also for example errno values could be used for errors.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/mgmt/osdp/src/osdp_cp.c (hunk 2 above)
Comment (cbsiddharth) [reply subsys/mgmt/osdp/src/osdp_cp.c:57]: Raised PR #27618. Will remove the one here once it is merged.

File: subsys/mgmt/osdp/src/osdp_cp.c (hunk 4 above)
Comment (cbsiddharth) [reply subsys/mgmt/osdp/src/osdp_cp.c]: I have fixed this to some extent (where ever there are more than 2 return codes). I think I must do the same to PD mode also. I will add a new commit that will do the same change there.

File: subsys/mgmt/osdp/src/osdp_cp.c (hunk 3 above)
Comment (cbsiddharth) [reply subsys/mgmt/osdp/src/osdp_cp.c]: Sorry, force of habit. I'll try to be more careful. Btw, can we have checkpatch warn for this?

File: subsys/mgmt/osdp/src/osdp_cp.c (hunk 1 above)
Comment (cbsiddharth) [reply subsys/mgmt/osdp/src/osdp_cp.c]: Done.

**Review by jukkar — APPROVED** [review]
LGTM

File: subsys/mgmt/osdp/src/osdp_cp.c (hunk 2 above)
Comment (cbsiddharth) [reply subsys/mgmt/osdp/src/osdp_cp.c]: #27618 is merged. So osdp_strsep() is replaced by strtok_r().

**Comment by cbsiddharth** [comment]
@carlescufi, this is ready for review.

**Comment by cbsiddharth** [comment]
Ping?

@carlescufi @pfalcon @galak

**Review by carlescufi — APPROVED** [review]
